modules/programs/rofi.nix

```
@@ -104,26 +104,39 @@ let
     else
       { };
 
-  mkValueString = value:
+  mkValueString = { quoteStr }:
+    value:
     if isBool value then
       if value then "true" else "false"
     else if isInt value then
       toString value
+    else if quoteStr then
+      ''"${toString value}"''
```

I guess it is color strings that need to be unquoted? If so, then perhaps
```suggestion
    else if isString value && builtins.match "#[[:xdigit:]]{6}" != null then
      value
```
?

---

No, almost everything needs to be unquoted. The only example I found of quote is this one: https://github.com/davatorium/rofi-themes/blob/master/Official%20Themes/Adapta-Nokto.rasi#L140

---

I don't think there is a easy way to differentiate, so I expect that the user quote the string when needed. Maybe I should document it :thinking: ?

---

I add a example to make it clear that if you need to set a string you will need to add quotes (and escape them). Should be rare the cases you need, though (I am using this PR here and didn't need to set strings: https://github.com/thiagokokada/dotfiles/commit/a961f69860016df96e6f88deb8a5f3b04c945faf)

---

I see. Hmm, it's quite unfortunate that there will be two separate representations of the rasi format within the same module. I think it would be good if they could be unified so that people don't get surprised. Perhaps something along the lines of the following patch could work?

```patch
diff --git a/modules/programs/rofi.nix b/modules/programs/rofi.nix
index 82099492..3c23f3e9 100644
--- a/modules/programs/rofi.nix
+++ b/modules/programs/rofi.nix
@@ -109,10 +109,12 @@ let
       if value then "true" else "false"
     else if isInt value then
       toString value
-    else if quoteStr then
-      ''"${toString value}"''
+    else if value._type or "" == "literal" then
+      value.value
+    else if isString value then
+      ''"${value}"''
     else
-      toString value;
+      abort "Unhandled value type ${builtins.typeOf value}";
 
   mkKeyValue = { quoteStr }:
     name: value:
@@ -150,7 +152,23 @@ let
     left = 8;
   };
 
-  configType = with types; attrsOf (oneOf [ str int bool ]);
+  configType = with types; attrsOf (oneOf [ str int bool rasiLiteral ]);
+
+  rasiLiteral = types.submodule {
+    options = {
+      _type = mkOption {
+        type = types.enum [ "literal" ];
+        internal = true;
+      };
+
+      value = mkOption {
+        type = types.str;
+        internal = true;
+      };
+    };
+  } // {
+    description = "Rasi literal string";
+  };
 
   themeType = with types; attrsOf configType;
 
@@ -372,6 +390,11 @@ in {
       '';
     }];
 
+    lib.formats.rasi.mkLiteral = value: {
+      _type = "literal";
+      inherit value;
+    };
+
     home.packages = [ cfg.package ];
 
     home.file."${cfg.configPath}".text = toRasi { } {
diff --git a/tests/modules/programs/rofi/custom-theme.nix b/tests/modules/programs/rofi/custom-theme.nix
index 6990c743..10dd63c9 100644
--- a/tests/modules/programs/rofi/custom-theme.nix
+++ b/tests/modules/programs/rofi/custom-theme.nix
@@ -7,10 +7,10 @@ with lib;
     programs.rofi = {
       enable = true;
 
-      theme = {
+      theme = with config.lib.formats.rasi; {
         "*" = {
-          background-color = "#000000";
-          border-color = "#FFFFFF";
+          background-color = mkLiteral "#000000";
+          border-color = mkLiteral "#FFFFFF";
           width = 512;
         };
```

---

I quite liked this way of declaring a special type (much better than my implementation), but I still not sure about exposing this type and forcing the user to call `mkLiteral`. I am more about refactoring the internal module but still setting `mkLiteral` for all values in themes, while configs will be called without it.

If you look at my usage here: https://github.com/thiagokokada/dotfiles/commit/a961f69860016df96e6f88deb8a5f3b04c945faf#diff-ff6693548a9dd5bc20d988c0d3a18b4058904a7653d22246f15ee065657fb839R22-R66, it will become quite clear that almost all values will need to be called with `mkLiteral` so this makes very ugly to declare a theme.

Also, rasi is only used by rofi AFAIK and there is only this two files that uses it (configuration and themes), so I don't think this will cause too much confusion either.

---

Actually, now that I think about it, users can always do something like this:

```
with libs; attrsets.mapAttrsRecursive (n: v:  if isString v then formats.rasi.mkLiteral v else v) { my: { theme = "something"; } }
```

And making it explicitly makes it more robust. So I decided to apply your changes.

---

Yeah, I imagine most people will end up doing something like that, or
```
theme = let l = config.lib.formats.rasi.mkLiteral; in {
  "*" = {
    background-color = l "#000000";
    border-color = l "#FFFFFF";
    width = 512;
  };
};
```